bin/cache.cpp

```diff
@@ -0,0 +1,97 @@
+#include <mbgl/storage/default_file_source.hpp>
+#include <mbgl/storage/resource.hpp>
+#include <mbgl/util/chrono.hpp>
+#include <mbgl/util/run_loop.hpp>
+
+#include <args.hxx>
+#include <mapbox/io.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+int main(int argc, char* argv[]) {
+    args::ArgumentParser p("Mapbox GL cache tool", "");
+    args::HelpFlag helpFlag(p, "help", "Display this help menu", {'h'});
+
+    args::ValueFlag<std::string> urlValue(p, "URL", "Resource URL (required)", {'u'}, args::Options::Required);
+    args::ValueFlag<std::string> cacheValue(
+        p, "cache", "Path to the cache database (required)", {'c'}, args::Options::Required);
+    args::ValueFlag<std::string> dataValue(
+        p, "data", "Path to the resource data (required)", {'d'}, args::Options::Required);
+    args::ValueFlag<std::string> etagValue(p, "etag", "Cache eTag, none otherwise", {'t'});
+    args::ValueFlag<unsigned long> expiresValue(p, "expires", "Expires date, will use 'now' otherwise", {'e'});
+    args::ValueFlag<unsigned long> modifiedValue(p, "modified", "Modified date, will use 'now' otherwise", {'m'});
+
+    std::unordered_map<std::string, mbgl::Resource::Kind> typeMap{{"glyphs", mbgl::Resource::Glyphs},
+                                                                  {"image", mbgl::Resource::Image},
+                                                                  {"source", mbgl::Resource::Source},
+                                                                  {"sprite-image", mbgl::Resource::SpriteImage},
+                                                                  {"sprite-json", mbgl::Resource::SpriteJSON},
+                                                                  {"style", mbgl::Resource::Style},
+                                                                  {"tile", mbgl::Resource::Tile}};
+
+    std::string typeHelp("One of the following (required):");
+    for (auto key : typeMap) {
+        typeHelp += " " + key.first;
+    }
+
+    args::MapFlag<std::string, mbgl::Resource::Kind> typeFlag(
+        p, "type", typeHelp, {"type"}, typeMap, args::Options::Required);
+
+    args::Group tileIdGroup(p, "Coordinates (required for 'tile')", args::Group::Validators::AllOrNone);
+    args::ValueFlag<int32_t> xValueFlag(tileIdGroup, "x", "Tile x coordinate", {'x'});
+    args::ValueFlag<int32_t> yValueFlag(tileIdGroup, "y", "Tile y coordinate", {'y'});
+    args::ValueFlag<int32_t> zValueFlag(tileIdGroup, "z", "The zoom level", {'z'});
```

Comment:
should these be unsigned?

Reply:
I thought the same first, but for some reason we use signed here:

```
    struct TileData {
        std::string urlTemplate;
        uint8_t pixelRatio;
        int32_t x;
        int32_t y;
        int8_t z;
    };
```